File: 2014-05-16-FP/FilterElements/fe.accepted.scala
import scala.collection.immutable.Map

object Solution {
    def comb(xs: List[Int]) = {
        def g(m: Map[Int, Int], x: Int) =
            (m get x) match {
                case None => m updated (x, 1)
                case Some(n) => m updated (x, n + 1)
            }
        (xs foldLeft (Map.empty[Int, Int]))(g)
    }

    def f0(xs: List[Int], k: Int): List[Int] = {
        def g(acc: (List[Int], Map[Int, Int]), x: Int) =
            (acc._2 get x) match {
                case None => acc
                case Some(n) =>
                    if (n >= k)
                        (x :: acc._1, acc._2 - x)
                    else
                        (acc._1, acc._2 - x)
            }
        (xs.foldLeft (List[Int](), comb(xs)))(g)._1.reverse
    }

    def f(xs: List[Int], k: Int) =
        f0(xs, k) match {
            case Nil => "-1"
            case xs => (xs map (_.toString)) mkString " "
        }

    def main(args: Array[String]) = {
        val t = readLine.toInt
        for (ix <- 0 until t) {
            val p = readLine.split(" ").map(_.toInt).toList
            val xs = readLine.split(" ").map(_.toInt).toList
            println(f(xs, p(1)))
        }
    }
}

File: 2014-05-16-FP/FilterElements/fe.accepted.fs
open System

let comb xs =
    let g m x =
        match (Map.tryFind x m) with
        | None -> Map.add x 1 m
        | Some n -> Map.add x (n + 1) m
    List.fold g Map.empty xs

let f0 xs k =
    let g ((acc, m) as ac) x =
        match (Map.tryFind x m) with
        | None -> ac
        | Some n ->
            if n >= k
                then (x :: acc, Map.remove x m)
                else (acc, Map.remove x m)
    List.rev (fst (List.fold g ([], comb xs) xs))

let f xs k =
    match (f0 xs k) with
    | [] -> "-1"
    | xs -> String.Join (" ", List.map string xs)

let tst _ =
    let p = List.map int (List.ofSeq ((Console.ReadLine ()).Split [|' '|]))
    let k = List.head (List.tail p)
    let xs = List.map int (List.ofSeq ((Console.ReadLine ()).Split [|' '|]))
    Console.WriteLine ("{0}", f xs k)

let main _ =
    let t = int (Console.ReadLine ())
    ignore (List.map tst [1 .. t])

main ()
